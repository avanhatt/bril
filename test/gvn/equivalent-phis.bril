// Ported from /llvm/test/Transforms/NewGVN/equivalent-phi.ll

// We should be able to prove the equivalence of two of the phis, and then use that to eliminate
// one set of indexing calculations and a load

// CHECK-LABEL: @bar(
// CHECK-NEXT:  bb:
// CHECK-NEXT:    br label bb3
// CHECK:       bb3:
// CHECK-NEXT:    [[TMP:.*]] = phi i32 [ arg, bb ], [ [[TMP:.*]]15, bb17 ]
// CHECK-NEXT:    [[TMP4:.*]] = phi i32 [ arg2, bb ], [ [[TMP18:.*]], bb17 ]
// CHECK-NEXT:    [[TMP6:.*]] = phi i32 [ 0, bb ], [ [[TMP14:.*]], bb17 ]
// CHECK-NEXT:    [[TMP7:.*]] = sext i32 [[TMP]] to i64
// CHECK-NEXT:    [[TMP8:.*]] = getelementptr inbounds [1024 x i32], [1024 x i32]* @global, i64 0, i64 [[TMP7]]
// CHECK-NEXT:    [[TMP9:.*]] = load i32, i32* [[TMP8]], align 4
// CHECK-NEXT:    [[TMP10:.*]] = add nsw i32 [[TMP6]], [[TMP9]]
// CHECK-NEXT:    [[TMP14]] = add nsw i32 [[TMP10]], [[TMP9]]
// CHECK-NEXT:    [[TMP15:.*]] = add nsw i32 [[TMP]], arg1
// CHECK-NEXT:    br label bb17
// CHECK:       bb17:
// CHECK-NEXT:    [[TMP18]] = add i32 [[TMP4]], -1
// CHECK-NEXT:    [[TMP19:.*]] = icmp ne i32 [[TMP4]], 0
// CHECK-NEXT:    br i1 [[TMP19]], label bb3, label bb20
// CHECK:       bb20:
// CHECK-NEXT:    ret i32 [[TMP14]]
//


int main(arg : int, arg1 : int, arg2 : int) {
bb:
  jmp bb3;

bb3:
  tmp : int = phi arg tmp15 bb bb17;
  tmp4 : int = phi arg2 tmp18 bb bb17;
  // Apparently we should be able to get rid of this!
  tmp5 : int = phi arg tmp16 bb bb17;
  zero : int = const 0;
  tmp6 : int = phi zero tmp14 bb bb17;
  tmp7 : int = id tmp;
  tmp8 : int = id tmp7;
  tmp9 : int = id tmp8;
  tmp10 : int = add tmp6 tmp9;
  tmp11 : int = id tmp5;
  tmp12 : int = id tmp11;
  tmp13 : int = id tmp12;
  tmp14 : int = add tmp10 tmp13;
  tmp15 : int = add tmp arg1;
  tmp16 : int = add tmp5 arg1;
  jmp bb17;

bb17:
  neg1 : int = const -1;
  tmp18 : int = add tmp4 neg1;
  eq : bool = eq tm4 zero;
  tmp19 : int = not eq;
  br tmp19 bb3 bb20;

bb20:
  ret tmp14;
}