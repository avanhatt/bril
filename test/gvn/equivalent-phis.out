int main(arg : int, arg1 : int, arg2 : int) {
bb:
  jmp copy1;
bb3:
  zero1: int = const 0;
  tmp151: int = add tmp1 arg1;
  tmp161: int = add tmp51 arg1;
  jmp bb17;
bb17:
  neg11: int = const -1;
  tmp181: int = add tmp41 neg11;
  eq1: bool = eq tm4 zero1;
  tmp191: int = not eq1;
  br tmp191 copy2 bb20;
bb20:
  ret tmp141;
copy1:
  tmp1: int = id arg;
  tmp41: int = id arg2;
  tmp51: int = id arg;
  jmp bb3;
copy2:
  tmp1: int = id tmp151;
  tmp41: int = id tmp181;
  tmp51: int = id tmp161;
  jmp bb3;
}
